socket.go

```diff
@@ -561,123 +613,255 @@ func (socket *mongoSocket) Query(ops ...interface{}) (err error) {
+			if ok {
+				delete(socket.replyFuncs, uint32(responseTo))
+			}
+			socket.Unlock()
+
+			if replyFunc != nil && reply.replyDocs == 0 {
+				replyFunc(nil, &reply, -1, nil)
+			} else {
+				for i := 0; i != int(reply.replyDocs); i++ {
+					_, err := io.ReadFull(r, s)
+					if err != nil {
+						if replyFunc != nil {
+							replyFunc(err, nil, -1, nil)
+						}
+						socket.kill(err, true)
+						return
+					}
+					b := make([]byte, int(getInt32(s, 0)))
+					copy(b[0:4], s)
+
+					_, err = io.ReadFull(r, b[4:])
+					if err != nil {
+						if replyFunc != nil {
+							replyFunc(err, nil, -1, nil)
+						}
+						socket.kill(err, true)
+						return
 					}
-					socket.kill(err, true)
-					return
-				}
-
-				b := make([]byte, int(getInt32(s, 0)))
 
-				// copy(b, s) in an efficient way.
-				b[0] = s[0]
-				b[1] = s[1]
-				b[2] = s[2]
-				b[3] = s[3]
+					if globalDebug && globalLogger != nil {
+						m := bson.M{}
+						if err := bson.Unmarshal(b, m); err == nil {
+							debugf("Socket %p to %s: received document: %#v", socket, socket.addr, m)
+						}
+					}
 
-				err = fill(conn, b[4:])
-				if err != nil {
 					if replyFunc != nil {
-						replyFunc(err, nil, -1, nil)
+						replyFunc(nil, &reply, i, b)
 					}
-					socket.kill(err, true)
-					return
+					// XXX Do bound checking against totalLen.
 				}
+			}
 
-				if globalDebug && globalLogger != nil {
-					m := bson.M{}
-					if err := bson.Unmarshal(b, m); err == nil {
-						debugf("Socket %p to %s: received document: %#v", socket, socket.addr, m)
-					}
-				}
+			socket.Lock()
+			if len(socket.replyFuncs) == 0 {
+				// Nothing else to read for now. Disable deadline.
+				socket.conn.SetReadDeadline(time.Time{})
+			} else {
+				socket.updateDeadline(readDeadline)
+			}
+			socket.Unlock()
 
-				if replyFunc != nil {
-					replyFunc(nil, &reply, i, b)
-				}
+		case dbMessage:
+			body := getEmptyBody(int(totalLen) - 16)
+			_, err := io.ReadFull(r, body)
+			if err != nil {
+				socket.kill(err, true)
+				return
+			}
+
+			sections, err := getSections(body[4:])
+			if err != nil {
+				socket.kill(err, true)
+				return
+			}
+			// TODO check CRC-32 checksum if checksum byte is set
+			reply := &msgOp{
+				flags:    uint32(getInt32(body, 0)),
+				sections: sections,
+			}
 
-				// XXX Do bound checking against totalLen.
+			// TODO update this when msgPayload1 section is implemented in MongoDB
+			stats.receivedDocs(1)
+			socket.Lock()
+			opMsgReplyFunc, ok := socket.opMsgReplyFuncs[uint32(responseTo)]
+			if ok {
+				delete(socket.opMsgReplyFuncs, uint32(responseTo))
 			}
-		}
+			socket.Unlock()
 
-		socket.Lock()
-		if len(socket.replyFuncs) == 0 {
-			// Nothing else to read for now. Disable deadline.
+			if opMsgReplyFunc != nil {
+				opMsgReplyFunc(reply, err)
+			} else {
+				socket.kill(fmt.Errorf("couldn't handle response properly"), true)
+				return
+			}
+			responsePool.Put(body)
```

Comment:
`defer` this on line 821 - currently this leaks the buffer on lines 824, 830, and 851.

I know some people dislike `defer` because it _used to be_ slow - the cost is ~0.00004ms these days (~40ns - I benchmarked recently on 1.9.2). The maintenance cost is well worth the 40ns in my opinion.

Reply:
Typically though I would recommend calling `pool.Get()` in the same func as `pool.Put()`and passing the slice into `getEmptyBody()`:

```go
var buffer = pool.Get().([]byte)[:0]
var body = getEmptyBody(buffer)
defer pool.Put(body)
```

Where `getEmptyBody()` calls `append()` on buffer, and returns the appended slice (which is backed by the same array):

```go
func getEmptyBody(buffer) {
    var newBuf = append(buffer, something)
    // etc...

    return newBuf
}
```

This keeps all the pooling in the same code block and easy to read, making a leak less likely when refactoring. I also always reslice on the `Get()` call so you're sure you've never got a dirty buffer, but I know the existing code does it on `Put()` - I'll sort it all one day!

Reply:
totally agree with you, this is an excellent idea ! I've just pushed a commit implementing this